dev.exs

```diff
@@ -28,6 +28,88 @@ Application.put_env(:phoenix_live_dashboard, DemoWeb.Endpoint,
   ]
 )
 
+defmodule DemoWeb.History do
+  use GenServer
+  alias Phoenix.LiveDashboard.TelemetryListener
+
+  @history_buffer_size 50
+
+  def data(metric) do
+    GenServer.call(__MODULE__, {:data, metric})
+  end
+
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  def init(metrics) do
+    GenServer.cast(__MODULE__, {:metrics, metrics})
+    {:ok, %{}}
+  end
+
+  def handle_cast({:metrics, metrics}, _state) do
+    {:noreply,
+     for metric <- metrics, reduce: %{} do
+       acc ->
+         key_metrics = Map.get(acc, event(metric.name), [])
+         metric_map = %{metric: metric, history: CircularBuffer.new(@history_buffer_size)}
+         attach_handler(metric, length(key_metrics))
+
+         Map.merge(acc, %{event(metric.name) => [metric_map | key_metrics]})
+     end}
+  end
+
+  defp attach_handler(%{name: name_list} = metric, id) do
+    :telemetry.attach(
+      "#{inspect(name_list)}-history-#{id}",
+      event(name_list),
+      &__MODULE__.handle_event/4,
+      metric
+    )
+  end
+
+  defp event(name_list) do
+    Enum.slice(name_list, 0, length(name_list) - 1)
+  end
+
+  def handle_event(event_name, data, metadata, metric) do
+    GenServer.cast(__MODULE__, {:telemetry_metric, event_name, data, metadata, metric})
+  end
+
+  def handle_cast({:telemetry_metric, event_name, data, metadata, metric}, state) do
+    if histories_list = state[event_name] do
+      time = System.system_time(:microsecond)
+
+      {%{history: history}, index} =
+        histories_list
+        |> Enum.with_index()
+        |> Enum.find(fn {map, _index} -> map.metric == metric end)
+
+      measurement = TelemetryListener.extract_measurement(metric, data)
+      label = TelemetryListener.tags_to_label(metric, metadata)
```

Comment:
Compute those in `handle_event` instead and send them on cast. You want to avoid passing the metadata around as much as possible.